Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+    private let _impl: PathImpl
+    
+    /// Initializes the AbsolutePath from `string`, which must be an absolute
+    /// path (which means that it must begin with either a path separator or a
+    /// tilde, denoting an absolute path of a user account home directory).
+    ///
+    /// The path string will be normalized by:
+    /// - Collapsing `..` path components
+    /// - Removing `.` path components
+    /// - Expanding `~` or `~user` prefix
+    /// - Removing any trailing path separator
+    /// - Removing any redundant path separators
+    ///
+    /// This string manipulation may change the meaning of a path if any of the
+    /// path components are symbolic links on disk.  However, the file system is
+    /// never accessed in any way when initializing an AbsolutePath.
+    public init(_ string: String) {
+        precondition(string.characters.first == PathSeparatorCharacter
+                  || string.characters.first == HomeDirectoryCharacter)
+        
+        // Get a hold of the character view.
+        // FIXME: We need to investigate what is most performant here.
+        //        For example, is UTF-8 view more efficient, or unichars, etc?
+        var chars = string.characters
+        
+        // Expand any `~` or `~user` prefix.
+        if chars.first == HomeDirectoryCharacter {
+            // We're doing home directory substitution. Check for a user name.
+            let nStart = chars.index(after: chars.startIndex)
+            let nEnd = chars.index(of: PathSeparatorCharacter) ?? chars.endIndex
+            if nStart == nEnd {
+                // No user name, so we need home directory of effective user.
+                let homeDir = gethomedir().characters
+                chars.replaceSubrange(chars.startIndex ..< nEnd, with: homeDir)
+            }
+            else {
+                // A user name was provided, so we try to look that up.
+                // FIXME:  What do we do if we cannot find the user name? Most
+                //         shells leave it as a relative path, which we do not
+                //         want to do since we want the type safety of knowing
+                //         up-front whether a path is absolute or relative.
+                let name = String(chars[nStart ..< nEnd])
+                let homeDir = gethomedir(user: name).characters
+                chars.replaceSubrange(chars.startIndex ..< nEnd, with: homeDir)
+            }
+        }
+        
+        // At this point we expect to have a path separator as first character.
+        assert(chars.first == PathSeparatorCharacter)
+
+        // Split the character array into parts, folding components as we go.
+        // As we do so, we count the number of characters we'll end up with in
+        // the normalized string representation.
+        var parts: [String.CharacterView] = []
+        var capacity = 0
+        for part in chars.split(separator: PathSeparatorCharacter) {
```

Comment:
Splitting on each component, building an array of parts, and reassembling is going to be somewhat expensive. It would be nice if we had a fast path here for cases where the input string is already normalized, can we get a FIXME for that (or the fast path, if it is easy)?